pyrate/core/phase_closure/mst_closure.py

```diff
@@ -82,29 +84,40 @@ def edges(self):
         return [Edge(swe.first, swe.edge.second) for swe in self.loop]
 
 
-def __discard_cycles_with_same_members(simple_cycles: List[List[date]]) -> List[List[date]]:
-    seen_sc_sets = set()
-    filtered_sc = []
-    for sc in simple_cycles:
-        loop = sc[:]
-        sc.sort()
-        sc = tuple(sc)
-        if sc not in seen_sc_sets:
-            seen_sc_sets.add(sc)
-            filtered_sc.append(loop)
-    return filtered_sc
+def __discard_cycles_with_same_members(cycles: List[List[date]]) -> List[List[date]]:
+    log.debug(f"Discarding loops that contain the same set of edges as another loop")
+    seen_cycle_sets = set()
+    filtered_cycles = []
+    for c in cycles:
+        loop = c[:]
+        c.sort()
+        c = tuple(c)
+        if c not in seen_cycle_sets:
+            seen_cycle_sets.add(c)
+            filtered_cycles.append(loop)
+    log.debug(f"Number of remaining loops is {len(filtered_cycles)}")
+    return filtered_cycles
 
 
-def __find_closed_loops(edges: List[Edge]) -> List[List[date]]:
+def __find_closed_loops(edges: List[Edge], max_loop_length: int) -> List[List[date]]:
     g = nx.Graph()
     edges = [(we.first, we.second) for we in edges]
     g.add_edges_from(edges)
     dg = nx.DiGraph(g)
-    simple_cycles = nx.simple_cycles(dg)  # will have all edges
-    simple_cycles = [scc for scc in simple_cycles if len(scc) > 2]  # discard edges
+    log.debug(f"Evaluating all possible loops using NetworkX simple_cycles function")
+    simple_cycles = nx.simple_cycles(dg)  # yields a generator that determines all possible cycles
+    log.debug(f"Total number of possible loops is {len(list(simple_cycles))}")
+    log.debug(f"Discarding loops with less than 3 edges and more than {max_loop_length} edges")
+    loop_subset = [scc for scc in simple_cycles
+                     if
+                     (len(scc) > 2)  # three or more edges reqd for closed loop
+                     and
+                     (len(scc) <= max_loop_length)  # discard loops exceeding max loop length
+                     ]
```

Comment:
Here is where the problem lies.
i) For large ifg networks (i.e. ~100 ifgs or greater), this is the section that is hanging (identified by the printing of debug statements).
ii) For the cropA data of 30 ifgs, see the output of the algorithm as follows:
`21:18:01 closure_check 135 3202011 INFO 0/0 Performing closure check on original set of 30 ifgs
21:18:02 mst_closure 107 3202011 DEBUG 0/0 Evaluating all possible loops using NetworkX simple_cycles function
21:18:02 mst_closure 109 3202011 DEBUG 0/0 Total number of possible loops is 5498
21:18:02 mst_closure 110 3202011 DEBUG 0/0 Discarding loops with less than 3 edges and more than 4 edges
21:18:02 mst_closure 117 3202011 DEBUG 0/0 Number of remaining loops is 0
21:18:02 mst_closure 88 3202011 DEBUG 0/0 Discarding loops that contain the same set of edges as another loop
21:18:02 mst_closure 98 3202011 DEBUG 0/0 Number of remaining loops is 0
21:18:02 closure_check 197 3202011 INFO 0/0 Total number of possible closure loops is 0
Traceback (most recent call last):
  File "/home/547/mcg547/PyRateVenv/bin/pyrate", line 33, in <module>
    sys.exit(load_entry_point('Py-Rate', 'console_scripts', 'pyrate')())
  File "/home/547/mcg547/repo/PyRate/pyrate/main.py", line 157, in main
    correct.main(config)
  File "/home/547/mcg547/repo/PyRate/pyrate/correct.py", line 135, in main
    return correct_ifgs(config)
  File "/home/547/mcg547/repo/PyRate/pyrate/correct.py", line 208, in correct_ifgs
    correct_steps[step](params, config)
  File "/home/547/mcg547/repo/PyRate/pyrate/correct.py", line 152, in update_params_with_closure_checked_ifg_list
    ifg_files, ifgs_breach_count, num_occurences_each_ifg = filter_to_closure_checked_ifgs(config)
TypeError: cannot unpack non-iterable NoneType object
`

Reply:
ii) was my bad - and @basaks resolved it with 3fb6552

i) is still an issue. The line causing issue is `simple_cycles = list(nx.simple_cycles(dg))` where the generator provided by `nx.simple_cycles` is being evaluated by the `list` function. In my tests, 30 ifgs gives a possible loops number of 5498. 50 ifgs gives 864868, and takes 40 secs to evaluate on the above stated line. I haven't waited long enough to get an answer for 100 ifgs. And the networks we want to use have between 400-500 ifgs...

Reply:
resolved by 3d51058